doc/decisions/3_decided/internal_cache.md

```diff
@@ -73,14 +73,46 @@ Improve documentation to make people more aware of these two problems:
 
 Upon returning from `kdbGet`/`kdbSet`, set the keyname of parentKey to the key that actually is parent of the data that was loaded.
 I.e. to the mountpoint of the backend that contains parentKey.
+`parentKey` is already an inout-type argument, since we use both it's value and metadata to return some information.
 
-> @markus2330 and @atmaxinger find this behavior very unexpected
+This behaviour was found very unexpected by @markus2330 and @atmaxinger.
 
 "Fewer keys":
 
-IMO this is simply a bug in the "nothing changed" logic.
-I don't think we need some complicated COW scheme to solve this.
-I haven't had a chance to check this yet, but I think it is just a matter of copying the keys from `backendData->keys`, so they are actually there and we don't just assume they are there.
+@kodebach considers this simply a bug in the "nothing changed" logic.
+A partial solution would be to copy the keys from `backendData->keys`, so they are actually there, and we don't just assume they are there.
+Still some extra steps are required to make this work:
+
+```c
+TEST_F (Simple, NothingToDo2)
+{
+	using namespace kdb;
+	KDB kdb;
+	KeySet ks;
+
+	auto parentKey = "system:" + testRoot;
+
+	EXPECT_EQ (kdb.get (ks, parentKey), 0) << "nothing to do in get";
+
+	ks.append (Key (parentKey + "a", KEY_VALUE, "x", KEY_END));
+	kdb.set (ks, parentKey);
+
+	EXPECT_EQ (kdb.get (ks, parentKey), 0) << "nothing to do in get";
+	EXPECT_TRUE (ks.lookup (parentKey + "a")) << "key a not found";
+	EXPECT_EQ (ks.lookup (parentKey + "a").get<std::string> (), "x") << "key a with wrong value";
+	// TODO: adding the line below breaks the test
+	//	 This is because the Key instance is shared between `ks` and the internal data of `kdb`.
+	// ks.lookup(parentKey + "a").set("y");
+
+	KeySet ks2;
+	EXPECT_EQ (kdb.get (ks2, parentKey), 0) << "nothing to do in get";
+	EXPECT_TRUE (ks2.lookup (parentKey + "a")) << "key a not found";
+	EXPECT_EQ (ks2.lookup (parentKey + "a").get<std::string> (), "x") << "key a with wrong value";
+}
+```
+
+A very simple way to make it work would be to make the keys returned by kdbGet read-only.
```

Comment:
This should be a whole decision on its own. 
There might be usecases where a read-only key name wouldn't work.
Also, how do we define "initial creation" if `keyAddName` is still allowed?

Reply:
> E.g. for the Key name, I am quite in favor to make it immutable after initial creation (which might consist of several `keyAddName`).

See #4683 for a way of decoupling the keyname operations from the key API. However, I think we should absolutely avoid the word "immutable", because that is something impossible in C unless you mean something like `mprotect`.

> This should be a whole decision on its own.

Definitely, and not a high priority one, if we don't use it here.